python/dgl/sampling/pinsage.py

```diff
@@ -0,0 +1,214 @@
+"""PinSAGE sampler & related functions and classes"""
+
+import numpy as np
+
+from .. import backend as F
+from .. import convert
+from .. import transform
+from .. import function as fn
+from .randomwalks import random_walk
+from .neighbor import select_topk
+from ..base import EID
+
+
+class GenericPinSAGESampler(object):
+    """PinSAGE-like sampler extended to any heterographs, given a metapath.
+
+    Given a heterogeneous graph, PinSAGE neighbor sampler would generate a homogeneous
+    graph where the neighbors of each node are the most commonly visited nodes of the
+    same type by random walk with restarts.  The random walk with restarts are based
+    on a given metapath, which should have the same beginning and ending node type.
+
+    Parameters
+    ----------
+    G : DGLHeteroGraph
+        The bidirectional bipartite graph.
+
+        The graph should only have two node types: ``ntype`` and ``other_type``.
+        The graph should only have two edge types, one connecting from ``ntype`` to
+        ``other_type``, and another connecting from ``other_type`` to ``ntype``.
+
+        PinSAGE works on a bidirectional bipartite graph where for each edge
+        going from node u to node v, there exists an edge going from node v to node u.
+    ntype : str
+        The node type for which the graph would be constructed on.
+    metapath : list[str]
+        The metapath.
+    random_walk_length : int
+        The maximum number of steps of random walk with restarts.
+
+        Note that here we consider traversing from ``ntype`` to ``other_type`` then back
+        to ``ntype`` as a single step (i.e. a single step consists of two hops).
+
+        Usually considered a hyperparameter.
```

Comment:
Remove this argument? We could let the `metapath` argument have repetition.

Reply:
Actually, `random_walk_length` controls where the random walks would restart.  In PinSAGE, random walks can only restart when a metapath traversal ends (I do need to change the docstring of `random_walk_restart_prob` though).  So I don't think I can remove the `random_walk_length` argument.